src/services_models/services_constructor.jl

```diff
@@ -15,96 +44,182 @@ end
 function construct_services!(
     container::OptimizationContainer,
     sys::PSY.System,
+    stage::ArgumentConstructStage,
     services_template::ServicesModelContainer,
     devices_template::DevicesModelContainer,
 )
     isempty(services_template) && return
     incompatible_device_types = get_incompatible_devices(devices_template)
 
-    function _construct_valid_services!(service_model::ServiceModel)
-        @debug "Building $(get_component_type(service_model)) with $(get_formulation(service_model)) formulation"
-        services = get_component_type(service_model)[]
-        if validate_services!(
-            get_component_type(service_model),
-            services,
-            incompatible_device_types,
+    groupservice = nothing
+
+    for (key, service_model) in services_template
+        if get_formulation(service_model) === GroupReserve  # group service needs to be constructed last
+            groupservice = key
+            continue
+        end
+        !validate_service!(service_model, incompatible_device_types, sys) && continue
+        construct_service!(
+            container,
             sys,
+            stage,
+            service_model,
+            devices_template,
+            incompatible_device_types,
         )
-            construct_service!(
-                container,
-                services,
-                sys,
-                service_model,
-                devices_template,
-                incompatible_device_types,
-            )
-        end
     end
+    groupservice === nothing || construct_service!(
+        container,
+        sys,
+        stage,
+        services_template[groupservice],
+        devices_template,
+        incompatible_device_types,
+    )
+    return
+end
 
-    groupservice = nothing
+function construct_services!(
```

Comment:
We should be passing the model into the construct services call now. We don't need the template. I am confused about this function use.

Reply:
the template is used to figure out the incompatible devices, this is existing code logic I followed